**Author** : Joel Da Costa

**Student ID:** w2154055

**Peer reviewer(s):** Rodrigo Rosas Guzman, Sergio Llopis Mas

**Peer review date:** 09/02/2026

**Code Feedback**: Catch up to date before interim submission on Friday 13th
Ensure that the code implemented identifies the criteria of the tasks' 3 requirements
Adopt the use of the correct technical terms for the dataset given in the lectures.

**All code has been imported from Code Reuse Session 1 (Week 3), there is no external code usage.**

Importing Pandas as well as controlling rows and columns (Code Reuse Session 1 (Week 3) snippet 1, 18)

In [ ]:
#Import pandas library and asign it pd
import pandas as pd

#Set the maximum rows displayed
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 150)

Load our dataset (Code Reuse
Session 1 (Week 3) snippet 2)

In [ ]:
#Load dataset with pandas for data analysis
data = pd.read_csv('/content/loan_applications_data.csv')

/tmp/ipykernel_12579/435885360.py:2: DtypeWarning:

Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.



Display an sample of the first 10 rows (Code Reuse Session 1 (Week 3) snippet 3)

In [ ]:
# Method to return the first 10 rows of the dataset
data.head(10)

,id,applicant_account_No.,bank_sort_code,age,sex,education_qualifications,income,home_ownership,employment_length,loan_intent,loan_amount,loan_interest_rate,loan_income_ratio,payment_default_on_file,credit_history_length,loan_approval_status,max_loan_amount(allowed),credit_application_acceptance
0,717474,62012825,90025,21.0,M,Unknown,12000.0,OWN,0.0,EDUCATION,15000,6.99,0.12,N,4,Approved,-2426900,0.0
1,211588,3829308,239285,21.0,M,Unknown,13200.0,OWN,2.0,EDUCATION,25000,16.77,0.19,Y,3,Approved,-111739,0.0
2,597607,79622422,300083,23.0,M,Unknown,9600.0,RENT,5.0,MEDICAL,30000,12.42,0.31,N,3,Approved,-89000,0.0
3,684154,86043250,405179,40.0,M,Unknown,182004.0,RENT,3.0,EDUCATION,35000,8.00,0.19,N,11,Approved,35000,0.0
4,343369,73510088,300066,40.0,M,Unknown,90000.0,MORTGAGE,3.0,HOMEIMPROVEMENT,35000,12.42,0.39,N,14,Approved,35000,0.0
5,803527,78040195,608312,40.0,M,Unknown,131004.0,MORTGAGE,9.0,VENTURE,30000,7.90,0.23,N,13,Approved,30000,0.0
6,142876,28963792,236078,40.0,M,Unknown,150000.0,MORTGAGE,9.0,DEBTCONSOLIDATION,25000,11.89,0.17,N,11,Approved,25000,0.0
7,146399,69308685,405047,40.0,M,Unknown,132000.0,MORTGAGE,3.0,HOMEIMPROVEMENT,25000,16.82,0.22,N,17,Approved,25000,0.0
8,593612,56095425,165501,40.0,M,Unknown,80000.0,MORTGAGE,2.0,DEBTCONSOLIDATION,25000,10.25,0.31,N,11,Approved,25000,0.0
9,620220,54207305,700603,40.0,M,Unknown,80000.0,MORTGAGE,9.0,VENTURE,25000,18.67,0.30,N,17,Approved,25000,0.0


List the variable names frmo this data set (Code Reuse Session 1 (Week 3) snippet 4)

In [ ]:
# Return a list of all variables in the dataset
list(data.columns)

['id',
 'applicant_account_No.',
 'bank_sort_code',
 'age',
 'sex',
 'education_qualifications',
 'income',
 'home_ownership',
 'employment_length',
 'loan_intent',
 'loan_amount',
 'loan_interest_rate',
 'loan_income_ratio',
 'payment_default_on_file',
 'credit_history_length',
 'loan_approval_status',
 'max_loan_amount(allowed)',
 'credit_application_acceptance']

Return all data types for variables used (Code Reuse Session 1 (Week 3) snippet 5)

In [ ]:
# Returns all the data sets for all variables in this dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58645 entries, 0 to 58644
Data columns (total 18 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   id                             58645 non-null  int64  
 1   applicant_account_No.          58645 non-null  int64  
 2   bank_sort_code                 58645 non-null  int64  
 3   age                            58639 non-null  float64
 4   sex                            362 non-null    object 
 5   education_qualifications       58645 non-null  object 
 6   income                         58614 non-null  float64
 7   home_ownership                 58645 non-null  object 
 8   employment_length              58630 non-null  float64
 9   loan_intent                    58645 non-null  object 
 10  loan_amount                    58645 non-null  int64  
 11  loan_interest_rate             58635 non-null  float64
 12  loan_income_ratio              58645 non-null 

Method to only calculate (count) and display the dimensions of the dataset (Code Reuse Session 1 (Week 3) snippet 6)

In [ ]:

# Display the dimensions of the dataset , its instances and variables
data.shape

(58645, 18)

Show the descriptive values (Code Reuse Session 1 (Week 3) snippet 7)

In [ ]:
# Display the descriptive values for the numeric varibales in the dataset
data.describe()

,id,applicant_account_No.,bank_sort_code,age,income,employment_length,loan_amount,loan_interest_rate,loan_income_ratio,credit_history_length,max_loan_amount(allowed),credit_application_acceptance
count,58645.000000,5.864500e+04,58645.000000,58639.000000,5.861400e+04,58630.000000,58645.000000,58635.000000,58645.000000,58645.000000,5.864500e+04,58644.000000
mean,549416.624537,5.055471e+07,353737.500077,27.555040,6.405295e+04,4.702746,9217.556518,10.689216,0.159238,5.813062,2.823538e+05,0.142385
std,260004.378682,2.864283e+07,227527.682002,6.099005,3.793834e+04,3.999875,5563.807384,3.257150,0.091692,4.027495,5.148478e+07,0.349447
min,100006.000000,1.000384e+06,10004.000000,-30.000000,4.200000e+03,0.000000,500.000000,-11.140000,0.000000,2.000000,-2.426900e+06,0.000000
25%,324881.000000,2.573109e+07,165810.000000,23.000000,4.200000e+04,2.000000,5000.000000,7.880000,0.090000,3.000000,3.800300e+04,0.000000
50%,548673.000000,5.057678e+07,301495.000000,26.000000,5.800000e+04,4.000000,8000.000000,10.750000,0.140000,4.000000,6.239200e+04,0.000000
75%,773531.000000,7.539393e+07,600168.000000,30.000000,7.566200e+04,7.000000,12000.000000,12.990000,0.210000,8.000000,9.271600e+04,0.000000
max,999996.000000,9.999746e+07,980000.000000,180.000000,1.900000e+06,150.000000,35000.000000,200.000000,0.830000,30.000000,1.246799e+10,1.000000


Show descriptive values for the categorical values (Code Reuse Session 1 (Week 3) snippet 8)

In [ ]:
# Show the descriptive vaalues for the categorcal variables in the dataset
data.describe(include='object')

,sex,education_qualifications,home_ownership,loan_intent,payment_default_on_file,loan_approval_status
count,362,58645,58645,58645,58640,58644
unique,4,6,4,6,4,8
top,M,Unknown,RENT,EDUCATION,N,Approved
freq,126,58303,30594,12271,49933,50210


Display the distinct value of a chosen variable (Code Reuse Session 1 (Week 3) snippet 9)


In [ ]:
# Display the unique values in the education_qualification varibale
data['education_qualifications'].unique()

array(['Unknown ', 'Unknown', 'Higher Education', 'College',
       'High School', 'Apprenticeship '], dtype=object)

In [ ]:
# # Display the unique values in the loan_approval_status varibale
data['loan_approval_status'].unique()

array(['Approved', 'APPROVED', 'APPROVED ', 'Declined ', 'DECLINED ',
       'Reject', nan, 'Accept', 'Accept '], dtype=object)

In [ ]:
# # Display the unique values in the payment_default_on_file varibale
data['payment_default_on_file'].unique()

array(['N', 'Y', nan, 'NO', 'YES'], dtype=object)

Rreturn a count of the rows with null variables as numbers and percentages (Code Reuse Session 1 (Week 3) snippet 10,11)

In [ ]:
# COunting and displaying the number of missing values per variable
data.isnull().sum()

,0
id,0
applicant_account_No.,0
bank_sort_code,0
age,6
sex,58283
education_qualifications,0
income,31
home_ownership,0
employment_length,15
loan_intent,0


In [ ]:
# COunting and displaying the number of missing values per variable as a percentage
data.isna().sum()/len(data)*100

,0
id,0.000000
applicant_account_No.,0.000000
bank_sort_code,0.000000
age,0.010231
sex,99.382727
education_qualifications,0.000000
income,0.052860
home_ownership,0.000000
employment_length,0.025578
loan_intent,0.000000


Removing uneeded variables that will not assist in modelling. (Code Reuse Session 1 (Week 3) snippet 12)

Reasoning:

id,applicant_account_No.,bank_sort_code - These varaibles are strictly used to identify users, leaving these in would create irrelevant noise to the model.

sex - This variable has 99.382727% of its values misssing completely at random(mcar), as such this varibale canot be imputed reliably so the only option would be to drop this variable.

education_qualifications - This variable has 58303 values set as 'unknown' , this many unusable values cannot contribute to anything useful in the modelling.

credit_application_acceptance - This variable is data leakage, it is returning an outcome that determines if the loan is accepted or not.

In [ ]:
# Dropping unnecassary variables in the dataset that will not be used for modelling
data = data.drop(columns=(['id','applicant_account_No.','bank_sort_code', 'sex', 'education_qualifications', 'credit_application_acceptance']))

After dropping the variables go over data exploration again (Code Reuse Session 1 (Week 3) snippet 3,6,4,5,7,8)

In [ ]:
# Method to return the first 10 rows of the dataset after dropping variables
data.head(10)

,age,income,home_ownership,employment_length,loan_intent,loan_amount,loan_interest_rate,loan_income_ratio,payment_default_on_file,credit_history_length,loan_approval_status,max_loan_amount(allowed)
0,21.0,12000.0,OWN,0.0,EDUCATION,15000,6.99,0.12,N,4,Approved,-2426900
1,21.0,13200.0,OWN,2.0,EDUCATION,25000,16.77,0.19,Y,3,Approved,-111739
2,23.0,9600.0,RENT,5.0,MEDICAL,30000,12.42,0.31,N,3,Approved,-89000
3,40.0,182004.0,RENT,3.0,EDUCATION,35000,8.00,0.19,N,11,Approved,35000
4,40.0,90000.0,MORTGAGE,3.0,HOMEIMPROVEMENT,35000,12.42,0.39,N,14,Approved,35000
5,40.0,131004.0,MORTGAGE,9.0,VENTURE,30000,7.90,0.23,N,13,Approved,30000
6,40.0,150000.0,MORTGAGE,9.0,DEBTCONSOLIDATION,25000,11.89,0.17,N,11,Approved,25000
7,40.0,132000.0,MORTGAGE,3.0,HOMEIMPROVEMENT,25000,16.82,0.22,N,17,Approved,25000
8,40.0,80000.0,MORTGAGE,2.0,DEBTCONSOLIDATION,25000,10.25,0.31,N,11,Approved,25000
9,40.0,80000.0,MORTGAGE,9.0,VENTURE,25000,18.67,0.30,N,17,Approved,25000


In [ ]:
# Display the dimensions of the dataset , its instances and variables after dropping variables
data.shape

(58645, 12)

In [ ]:
# Return a list of all variables in the dataset after dropping variables
list(data.columns)

['age',
 'income',
 'home_ownership',
 'employment_length',
 'loan_intent',
 'loan_amount',
 'loan_interest_rate',
 'loan_income_ratio',
 'payment_default_on_file',
 'credit_history_length',
 'loan_approval_status',
 'max_loan_amount(allowed)']

In [ ]:
# Returns all the data types for all variables in this dataset after dropping varibales
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58645 entries, 0 to 58644
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       58639 non-null  float64
 1   income                    58614 non-null  float64
 2   home_ownership            58645 non-null  object 
 3   employment_length         58630 non-null  float64
 4   loan_intent               58645 non-null  object 
 5   loan_amount               58645 non-null  int64  
 6   loan_interest_rate        58635 non-null  float64
 7   loan_income_ratio         58645 non-null  float64
 8   payment_default_on_file   58640 non-null  object 
 9   credit_history_length     58645 non-null  int64  
 10  loan_approval_status      58644 non-null  object 
 11  max_loan_amount(allowed)  58645 non-null  int64  
dtypes: float64(5), int64(3), object(4)
memory usage: 5.4+ MB


In [ ]:
# Display the descriptive values for the numeric varibales in the dataset after dropping variables
data.describe()

,age,income,employment_length,loan_amount,loan_interest_rate,loan_income_ratio,credit_history_length,max_loan_amount(allowed)
count,58639.000000,5.861400e+04,58630.000000,58645.000000,58635.000000,58645.000000,58645.000000,5.864500e+04
mean,27.555040,6.405295e+04,4.702746,9217.556518,10.689216,0.159238,5.813062,2.823538e+05
std,6.099005,3.793834e+04,3.999875,5563.807384,3.257150,0.091692,4.027495,5.148478e+07
min,-30.000000,4.200000e+03,0.000000,500.000000,-11.140000,0.000000,2.000000,-2.426900e+06
25%,23.000000,4.200000e+04,2.000000,5000.000000,7.880000,0.090000,3.000000,3.800300e+04
50%,26.000000,5.800000e+04,4.000000,8000.000000,10.750000,0.140000,4.000000,6.239200e+04
75%,30.000000,7.566200e+04,7.000000,12000.000000,12.990000,0.210000,8.000000,9.271600e+04
max,180.000000,1.900000e+06,150.000000,35000.000000,200.000000,0.830000,30.000000,1.246799e+10


In [ ]:
# Show the descriptive vaalues for the categorcal variables in the dataset after dropping variables
data.describe(include='object')

,home_ownership,loan_intent,payment_default_on_file,loan_approval_status
count,58645,58645,58640,58644
unique,4,6,4,8
top,RENT,EDUCATION,N,Approved
freq,30594,12271,49933,50210


New import to visulise the variables in the dataset (Code Reuse Session 1 (Week 3) snippet 13)

In [ ]:
# Import the plotly.express library and asign it px
import plotly.express as px

Use histograms to show the distrbution of 'income'(Code Reuse Session 1 (Week 3) snippet 14)

In [ ]:
# Use a histogrm to show the distribution if the variable 'loan_amount'
loan_amount_fig = px.histogram(data, x='loan_amount')
loan_amount_fig.show()


Use histograms to show the distrbution of 'loan_amount' (Code Reuse Session 1 (Week 3) snippet 14)

In [ ]:
# Use a histogrm to show the distribution of the variable 'income'
income_fig = px.histogram(data, x = 'income')
income_fig.show()

Using a bar chart for categorical varibables specifically. (Code Reuse Session 1 (Week 3) snippet 15)

In [ ]:
# Display a bar chart to show the distribution of variable 'loan_approval_status'
loan_status_fig = px.bar(data, x = 'loan_approval_status')
loan_status_fig.show()

Using the box plot to visually identify outliers. (Code Reuse Session 1 (Week 3) snippe 16)

In [ ]:
# Display the outliers of the 'employment_length' variable to discover outliers
employment_length_box_fig = px.box(data, x= 'employment_length')
employment_length_box_fig.show()

In [ ]:
# Display the outliers of the 'income' variable to discover outliers
income_box_fig = px.box(data, x='income')
income_box_fig.show()

In [ ]:
# Display the outliers of the 'loan_amount' variable to discover outliers
loan_amount_box_fig = px.box(data, x= 'loan_amount')
loan_amount_box_fig.show()

In [ ]:
# Display the outliers of the 'loan_interest_rate' variable to discover outliers
interest_rate_box_fig = px.box(data, x= 'loan_interest_rate')
interest_rate_box_fig.show()

This funcrtion detects outliers in any numeric variable based on the inter-quartile range test (Code Reuse Session 1 (Week 3) snippet 17)

In [ ]:
# Creating a function that can call the outliers of a specific numeric variable
def find_outliers_IQR(data):

  q1=data.quantile(0.25)

  q3=data.quantile(0.75)

  IQR=q3-q1

  outliers = data[((data<(q1-1.5*IQR))|(data> (q3+1.5*IQR)))]

  return outliers

Call the function and show outliers for age,income, loan_interest_rate, loan_amount,  max_loan_amount(allowed), loan_income_ratio , credit_history_length and employment_length (Code Reuse Session 1 (Week 3) snippet 19)

In [ ]:
# Function call of the age variable to display its outliers
outliers = find_outliers_IQR(data['age'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 2452


,age
1108,3.0
1813,47.0
2904,1.0
4306,180.0
8334,41.0
...,...
58602,43.0
58608,42.0
58613,55.0
58614,42.0


In [ ]:
# Function call of the income variable to display its outliers
outliers = find_outliers_IQR(data['income'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 2410


,income
3,182004.0
5,131004.0
6,150000.0
7,132000.0
13,186996.0
...,...
58640,1200000.0
58641,1200000.0
58642,1900000.0
58643,1839784.0


In [ ]:
# Function call of the loan_interest_rate variable to display its outliers
outliers = find_outliers_IQR(data['loan_interest_rate'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 43


,loan_interest_rate
213,50.60
446,35.80
727,21.36
741,21.21
783,21.74
928,80.80
1248,120.00
1478,23.06
1599,150.00
1741,76.80


In [ ]:
# Function call of the loan_amount variable to display its outliers
outliers = find_outliers_IQR(data['loan_amount'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 2045


,loan_amount
1,25000
2,30000
3,35000
4,35000
5,30000
...,...
58580,25000
58582,35000
58619,25000
58635,25000


In [ ]:
# Function call of the max_loan_amount(allowed) variable to display its outliers
outliers = find_outliers_IQR(data['max_loan_amount(allowed)'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 2234


,max_loan_amount(allowed)
0,-2426900
1,-111739
2,-89000
49814,12467989660
56415,174791
...,...
58640,1903960
58641,1909968
58642,2088491
58643,2291708


In [ ]:
# Function call of the loan_income_ratio variable to display its outliers
outliers = find_outliers_IQR(data['loan_income_ratio'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 1210


,loan_income_ratio
10,0.42
388,0.44
395,0.44
403,0.40
405,0.43
...,...
44728,0.40
45168,0.40
45306,0.41
47899,0.44


In [ ]:
# Function call of the credit_history_length variable to display its outliers
outliers = find_outliers_IQR(data['credit_history_length'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 1990


,credit_history_length
7,17
9,17
11,16
13,17
15,17
...,...
58583,16
58587,17
58601,16
58602,16


In [ ]:
# Function call of the employment_length variable to display its outliers
outliers = find_outliers_IQR(data['employment_length'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 1275


,employment_length
12,19.0
13,18.0
29,19.0
31,22.0
32,18.0
...,...
58563,18.0
58607,16.0
58611,16.0
58639,18.0


From the outliers we can drop specifc outliers (Code Reuse Session 1 (Week 3) snippet 28)

In [ ]:
# Filter out clients with age below 18
data=data[data.age >= 18]

In [ ]:
# Filter out clients with age above 110
data=data[data.age <= 110]

In [ ]:
# Filter out clients with employment_length above 70
data=data[data.employment_length <= 70]

In [ ]:
# Filter out clients with a negative loan_interest_rate
data=data[data.loan_interest_rate > 0]

In [ ]:
# Filter out clients with loan_interest_rate above 100
data=data[data.loan_interest_rate <= 100]

From the detected outliers there was a detection of indexes that were extreme outliers (negatives and impossible high amounts) so these will be removed.
(Code Reuse Session 1 (Week 3) snippet 20)

In [ ]:
# Indices 0, 1, 2 contain negative max_loan_amount values which are impossible
data.drop(data.index[[0,1,2]], inplace=True)

Show the descriptive values (Code Reuse Session 1 (Week 3) snippet 7)

In [ ]:
# Display descriptive statistics after outlier removal
data.describe()

,age,income,employment_length,loan_amount,loan_interest_rate,loan_income_ratio,credit_history_length,max_loan_amount(allowed)
count,58598.000000,5.856700e+04,58598.000000,58598.000000,58598.000000,58598.000000,58598.000000,5.859800e+04
mean,27.551555,6.406329e+04,4.696201,9215.402010,10.681325,0.159185,5.813765,2.825946e+05
std,6.024456,3.794129e+04,3.893329,5562.256645,3.074847,0.091654,4.028210,5.150542e+07
min,20.000000,4.200000e+03,0.000000,500.000000,5.420000,0.000000,2.000000,0.000000e+00
25%,23.000000,4.200000e+04,2.000000,5000.000000,7.880000,0.090000,3.000000,3.808400e+04
50%,26.000000,5.800000e+04,4.000000,8000.000000,10.750000,0.140000,4.000000,6.245400e+04
75%,30.000000,7.577050e+04,7.000000,12000.000000,12.990000,0.210000,8.000000,9.272150e+04
max,84.000000,1.900000e+06,41.000000,35000.000000,80.800000,0.830000,30.000000,1.246799e+10


Call the function and show outliers for max_loan_amount(allowed) (Code Reuse Session 1 (Week 3) snippet 19)

In [ ]:
# Run the IQR function on max_loan_amount(allowed) after removal to check remaining outliers
outliers = find_outliers_IQR(data['max_loan_amount(allowed)'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 2230


,max_loan_amount(allowed)
49814,12467989660
56414,174776
56415,174791
56416,174817
56417,174945
...,...
58640,1903960
58641,1909968
58642,2088491
58643,2291708


Filter the dataset to remove unrealistic maximum loan amount values (Code Reuse Session 1 (Week 3) snippet 28)

In [ ]:
# Filter out max_loan_amount values above 10 million
data=data[data['max_loan_amount(allowed)'] < 10000000]

Show the descriptive values (Code Reuse Session 1 (Week 3) snippet 7)

In [ ]:
# Display descriptive statistics after all outlier removal
data.describe()

,age,income,employment_length,loan_amount,loan_interest_rate,loan_income_ratio,credit_history_length,max_loan_amount(allowed)
count,58597.000000,5.856600e+04,58597.000000,58597.000000,58597.000000,58597.000000,58597.000000,5.859700e+04
mean,27.551615,6.406268e+04,4.696145,9215.456883,10.681385,0.159187,5.813830,6.982423e+04
std,6.024490,3.794133e+04,3.893338,5562.288247,3.074838,0.091654,4.028213,6.087671e+04
min,20.000000,4.200000e+03,0.000000,500.000000,5.420000,0.000000,2.000000,0.000000e+00
25%,23.000000,4.200000e+04,2.000000,5000.000000,7.880000,0.090000,3.000000,3.808400e+04
50%,26.000000,5.800000e+04,4.000000,8000.000000,10.750000,0.140000,4.000000,6.245300e+04
75%,30.000000,7.574400e+04,7.000000,12000.000000,12.990000,0.210000,8.000000,9.272000e+04
max,84.000000,1.900000e+06,41.000000,35000.000000,80.800000,0.830000,30.000000,2.638778e+06


Simple imputation with mean (Code Reuse Session 1 (Week 3) snippet 21)

In [ ]:
# Calculate the mean of the age variable
mean_age = data['age'].mean()
# Fill missing age values using the mean age value
data['age'].fillna(mean_age, inplace=True)

/tmp/ipykernel_12579/657823765.py:4: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [ ]:
# Calculate the mean of the income variable
mean_income = data['income'].mean()
# Fill missing income values using the mean income value
data['income'].fillna(mean_income, inplace=True)

/tmp/ipykernel_12579/4054745385.py:4: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [ ]:
# Calculate the mean of the employment_length variable
mean_employment_length = data['employment_length'].mean()
# Fill missing employment_length values using the mean employment_length value
data['employment_length'].fillna(mean_employment_length, inplace=True)

/tmp/ipykernel_12579/3563969167.py:4: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [ ]:
# Calculate the mean of the loan_interest_rate variable
mean_loan_interest_rate = data['loan_interest_rate'].mean()
# Fill missing loan_interest_rate values using the mean loan_interest_rate value
data['loan_interest_rate'].fillna(mean_loan_interest_rate, inplace=True)

/tmp/ipykernel_12579/3159303428.py:4: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





Show the descriptive values (Code Reuse Session 1 (Week 3) snippet 7)

In [ ]:
# Display descriptive statistics after applying mean imputation
data.describe()

,age,income,employment_length,loan_amount,loan_interest_rate,loan_income_ratio,credit_history_length,max_loan_amount(allowed)
count,58597.000000,5.859700e+04,58597.000000,58597.000000,58597.000000,58597.000000,58597.000000,5.859700e+04
mean,27.551615,6.406268e+04,4.696145,9215.456883,10.681385,0.159187,5.813830,6.982423e+04
std,6.024490,3.793129e+04,3.893338,5562.288247,3.074838,0.091654,4.028213,6.087671e+04
min,20.000000,4.200000e+03,0.000000,500.000000,5.420000,0.000000,2.000000,0.000000e+00
25%,23.000000,4.200000e+04,2.000000,5000.000000,7.880000,0.090000,3.000000,3.808400e+04
50%,26.000000,5.800000e+04,4.000000,8000.000000,10.750000,0.140000,4.000000,6.245300e+04
75%,30.000000,7.563800e+04,7.000000,12000.000000,12.990000,0.210000,8.000000,9.272000e+04
max,84.000000,1.900000e+06,41.000000,35000.000000,80.800000,0.830000,30.000000,2.638778e+06


Formatiing catergorical values into numeric (Code Reuse Session 1 (Week 3) snippet 23)

In [ ]:
# Mapping the loan_approval_status labels into numeric values
data['loan_approval_status'] = data['loan_approval_status'].map({
    'Approved': 1, 'APPROVED': 1,'APPROVED ': 1, 'Accept ': 1,'Accept': 1,
    'Declined ': 0,'Declined': 0, 'Reject': 0, 'DECLINED ': 0,'DECLINED': 0
})

In [ ]:
# Mapping the payment_default_on_file labels into numeric values
data['payment_default_on_file'] = data['payment_default_on_file'].map({
    'N':0, 'NO':0,
    'Y':1, 'YES':1
})

In [ ]:
# Display descriptive statistics after mapping the binary categorical variables
data.describe()

,age,income,employment_length,loan_amount,loan_interest_rate,loan_income_ratio,payment_default_on_file,credit_history_length,loan_approval_status,max_loan_amount(allowed)
count,58597.000000,5.859700e+04,58597.000000,58597.000000,58597.000000,58597.000000,58592.000000,58597.000000,58596.000000,5.859700e+04
mean,27.551615,6.406268e+04,4.696145,9215.456883,10.681385,0.159187,0.148365,5.813830,0.857908,6.982423e+04
std,6.024490,3.793129e+04,3.893338,5562.288247,3.074838,0.091654,0.355464,4.028213,0.349147,6.087671e+04
min,20.000000,4.200000e+03,0.000000,500.000000,5.420000,0.000000,0.000000,2.000000,0.000000,0.000000e+00
25%,23.000000,4.200000e+04,2.000000,5000.000000,7.880000,0.090000,0.000000,3.000000,1.000000,3.808400e+04
50%,26.000000,5.800000e+04,4.000000,8000.000000,10.750000,0.140000,0.000000,4.000000,1.000000,6.245300e+04
75%,30.000000,7.563800e+04,7.000000,12000.000000,12.990000,0.210000,0.000000,8.000000,1.000000,9.272000e+04
max,84.000000,1.900000e+06,41.000000,35000.000000,80.800000,0.830000,1.000000,30.000000,1.000000,2.638778e+06


Display the distinct value of the target variables after imputation (Code Reuse Session 1 (Week 3) snippet 9)

In [ ]:
# Confirm the remaining number of missing values per variableb after mean imputation
data.isnull().sum()

,0
age,0
income,0
home_ownership,0
employment_length,0
loan_intent,0
loan_amount,0
loan_interest_rate,0
loan_income_ratio,0
payment_default_on_file,5
credit_history_length,0


In [ ]:
# Display the distinct values of payment_default_on_file after mapping
data['payment_default_on_file'].unique()

array([ 0.,  1., nan])

In [ ]:
# Display the distinct values of loan_approval_status after mapping

data['loan_approval_status'].unique()

array([ 1.,  0., nan])

In [ ]:
# Confirm the remaining number of missing values per variable
data.isnull().sum()

,0
age,0
income,0
home_ownership,0
employment_length,0
loan_intent,0
loan_amount,0
loan_interest_rate,0
loan_income_ratio,0
payment_default_on_file,5
credit_history_length,0


drop all the remaining null values(Code Reuse Session 1 (Week 3) snippet 22)

In [ ]:
# Drop the missing valued rows
data = data.dropna()

Final confirmation that there are no missing values (Code Reuse Session 1 (Week 3) snippet 10)

In [ ]:
# Confirm the remaining number of missing values per variable
data.isnull().sum()

,0
age,0
income,0
home_ownership,0
employment_length,0
loan_intent,0
loan_amount,0
loan_interest_rate,0
loan_income_ratio,0
payment_default_on_file,0
credit_history_length,0


Import skLearn preprocessing (Code Reuse Session 1 (Week 3) snippet 25)

In [ ]:
# Import the package from sklearn
from sklearn import preprocessing
label_encoder = preprocessing.LabelEncoder()

Encode the final 2 columns - 'home_ownership' , 'loan_intent' (Code Reuse Session 1 (Week 3) snippet 26)

In [ ]:
# Create a label encoder object to encode categorical variables
label_encoder = preprocessing.LabelEncoder()
# Encode the home_ownership variable into numeric labels
data['home_ownership'] = label_encoder.fit_transform(data['home_ownership'])
# Encode the loan_intent variable into numeric labels
data['loan_intent'] = label_encoder.fit_transform(data['loan_intent'])

Modify the datatype of these data types to integers (Code Reuse Session 1 (Week 3) snippet 24)

In [ ]:
# Convert the age and employment_length variables from float to integer type
data = data.astype({"age":"int", "employment_length":"int"})

Grouping the data into 2 sets of categorical and numeric(Code Reuse Session 1 (Week 3) snippet 29,30 )

In [ ]:
# Group the numeric variables into one list for scaling
numeric_col = ['age', 'income', 'employment_length', 'loan_amount', 'loan_interest_rate', 'loan_income_ratio', 'credit_history_length']
# Group the categorical variables into one list for organisation
catergorical_col = ['home_ownership', 'loan_intent', 'payment_default_on_file']

import MinMax scaler for numeric variable scaling (Code Reuse Session 1 (Week 3) snippet 33)

In [ ]:
# Import the MinMaxScaler class from sklearn
from sklearn.preprocessing import MinMaxScaler
# Create a MinMaxScaler object
scaler = MinMaxScaler()
# Apply minimum-maximum scaling to the numeric variables
data[numeric_col] = scaler.fit_transform(data[numeric_col])

final list of rows (Code Reuse Session 1 (Week 3) snippet 3)

In [ ]:
# Display the first five rows of the final cleaned dataset
data.head()

,age,income,home_ownership,employment_length,loan_intent,loan_amount,loan_interest_rate,loan_income_ratio,payment_default_on_file,credit_history_length,loan_approval_status,max_loan_amount(allowed)
3,0.3125,0.093788,3,0.073171,1,1.000000,0.034227,0.228916,0.0,0.321429,1.0,35000
4,0.3125,0.045258,0,0.073171,2,1.000000,0.092863,0.469880,0.0,0.428571,1.0,35000
5,0.3125,0.066887,0,0.219512,5,0.855072,0.032900,0.277108,0.0,0.392857,1.0,30000
6,0.3125,0.076907,0,0.219512,0,0.710145,0.085832,0.204819,0.0,0.321429,1.0,25000
7,0.3125,0.067412,0,0.073171,2,0.710145,0.151234,0.265060,0.0,0.535714,1.0,25000


Display the final data types and non-null counts after cleaning and scaling (Code Reuse Session 1 (Week 3) snippet 5)

In [ ]:
# Display the data types and non-null counts of the final cleaned dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 58591 entries, 3 to 58644
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       58591 non-null  float64
 1   income                    58591 non-null  float64
 2   home_ownership            58591 non-null  int64  
 3   employment_length         58591 non-null  float64
 4   loan_intent               58591 non-null  int64  
 5   loan_amount               58591 non-null  float64
 6   loan_interest_rate        58591 non-null  float64
 7   loan_income_ratio         58591 non-null  float64
 8   payment_default_on_file   58591 non-null  float64
 9   credit_history_length     58591 non-null  float64
 10  loan_approval_status      58591 non-null  float64
 11  max_loan_amount(allowed)  58591 non-null  int64  
dtypes: float64(9), int64(3)
memory usage: 5.8 MB


     Now to drop the targets for the two models and export the two cleaned datasets for ML modelling  (Code Reuse Session 1 (Week 3) snippet 12, 27, 28)

*   Regression
*   Classification



In [ ]:
# Create the classification dataset by dropping the regression target variable
classification = data.drop(columns=(['max_loan_amount(allowed)']))
# Create the regression dataset by keeping only approved clients
regression = data[data['loan_approval_status'] > 0]
# Drop the loan approval status target from the regression dataset
regression = regression.drop(columns=(['loan_approval_status']))

Display the first 10 rows of the prepared classification dataset (Code Reuse Session 1 (Week 3) snippet 3)

In [ ]:
# Display the first 10 rows of the classification dataset
classification.head(10)

,age,income,home_ownership,employment_length,loan_intent,loan_amount,loan_interest_rate,loan_income_ratio,payment_default_on_file,credit_history_length,loan_approval_status
3,0.3125,0.093788,3,0.073171,1,1.000000,0.034227,0.228916,0.0,0.321429,1.0
4,0.3125,0.045258,0,0.073171,2,1.000000,0.092863,0.469880,0.0,0.428571,1.0
5,0.3125,0.066887,0,0.219512,5,0.855072,0.032900,0.277108,0.0,0.392857,1.0
6,0.3125,0.076907,0,0.219512,0,0.710145,0.085832,0.204819,0.0,0.321429,1.0
7,0.3125,0.067412,0,0.073171,2,0.710145,0.151234,0.265060,0.0,0.535714,1.0
8,0.3125,0.039983,0,0.048780,0,0.710145,0.064075,0.373494,0.0,0.321429,1.0
9,0.3125,0.039983,0,0.219512,5,0.710145,0.175776,0.361446,0.0,0.535714,1.0
10,0.3125,0.029433,2,0.121951,5,0.710145,0.086893,0.506024,1.0,0.357143,1.0
11,0.3125,0.042621,0,0.073171,2,0.710145,0.089679,0.349398,0.0,0.500000,1.0
12,0.3125,0.034706,2,0.463415,1,0.710145,0.170204,0.433735,0.0,0.428571,1.0


Display the dimensions of the prepared classification dataset (Code Reuse Session 1 (Week 3) snippet 6)

In [ ]:
# Display the dimensions of the classification dataset
classification.shape

(58591, 11)

Display the variable names of the prepared classification dataset (Code Reuse Session 1 (Week 3) snippet 4)

In [ ]:
# Display the variable names of the classification dataset
list(classification.columns)

['age',
 'income',
 'home_ownership',
 'employment_length',
 'loan_intent',
 'loan_amount',
 'loan_interest_rate',
 'loan_income_ratio',
 'payment_default_on_file',
 'credit_history_length',
 'loan_approval_status']

Display the variable names of the prepared classification dataset (Code Reuse Session 1 (Week 3) snippet 4)

In [ ]:
# Display the first 10 rows of the regression dataset
regression.head(10)

,age,income,home_ownership,employment_length,loan_intent,loan_amount,loan_interest_rate,loan_income_ratio,payment_default_on_file,credit_history_length,max_loan_amount(allowed)
3,0.3125,0.093788,3,0.073171,1,1.000000,0.034227,0.228916,0.0,0.321429,35000
4,0.3125,0.045258,0,0.073171,2,1.000000,0.092863,0.469880,0.0,0.428571,35000
5,0.3125,0.066887,0,0.219512,5,0.855072,0.032900,0.277108,0.0,0.392857,30000
6,0.3125,0.076907,0,0.219512,0,0.710145,0.085832,0.204819,0.0,0.321429,25000
7,0.3125,0.067412,0,0.073171,2,0.710145,0.151234,0.265060,0.0,0.535714,25000
8,0.3125,0.039983,0,0.048780,0,0.710145,0.064075,0.373494,0.0,0.321429,25000
9,0.3125,0.039983,0,0.219512,5,0.710145,0.175776,0.361446,0.0,0.535714,25000
10,0.3125,0.029433,2,0.121951,5,0.710145,0.086893,0.506024,1.0,0.357143,25000
11,0.3125,0.042621,0,0.073171,2,0.710145,0.089679,0.349398,0.0,0.500000,25000
12,0.3125,0.034706,2,0.463415,1,0.710145,0.170204,0.433735,0.0,0.428571,25000


Display the dimensions of the prepared regression dataset (Code Reuse Session 1 (Week 3) snippet 6)

In [ ]:
# Display the dimensions of the regression dataset
regression.shape

(50266, 11)

Display the variable names of the prepared regression dataset (Code Reuse Session 1 (Week 3) snippet 4)

In [ ]:
# Display the variable names of the regression dataset
list(regression.columns)

['age',
 'income',
 'home_ownership',
 'employment_length',
 'loan_intent',
 'loan_amount',
 'loan_interest_rate',
 'loan_income_ratio',
 'payment_default_on_file',
 'credit_history_length',
 'max_loan_amount(allowed)']

Export the prepared classification and regression datasets to CSV files (Code Reuse Session 1 (Week 3) snippet 27)

In [ ]:
# Export the prepared classification dataset to a CSV file
classification.to_csv(r'Classification.csv', index=False)
# Export the prepared regression dataset to a CSV file
regression.to_csv(r'Regression.csv', index=False)

Display the dimensions of the final cleaned dataset for confirmation (Code Reuse Session 1 (Week 3) snippet 6)

In [ ]:
# Display the dimensions of the final cleaned dataset
data.shape

(58591, 12)